### PR TITLE
Add benchmarks for testing row filtering

### LINKED ISSUE #3457: Add benchmarks for parquet queries with filter pushdown enabled
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

We should add benchmarks for queries on parquet data with filter pushdown enabled (see https://github.com/apache/arrow-datafusion/pull/3380). 

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by thinkharderdev**
# Which issue does this PR close?

Closes #3457 

 # Rationale for this change

Need a set benchmarks for evaluating performance implications of parquet predicate pushdown. This sets up some very basic benchmarks which can be used for that purpose. Thanks to @tustvold for cooking up a script to generate synthetic datasets for this purpose. 

# What changes are included in this PR?

Add new benchmark script `parquet_filter_pushdown` which will execute a series of `ParquetExec` plans with different filter predicates. For each predicate in the suite, we will execute the plan with all three different `ParquetScanOptions` configurations:

1. Disable filter pushdown entirely (currently the default behavior)
2. Enable filter pushdown and predicate reordering
3. Enable filter pushdown without predicate reordering.  

# Are there any user-facing changes?

No

### CODE REVIEW

**Comment by thinkharderdev** [comment]
@alamb @tustvold @Ted-Jiang

File: benchmarks/src/bin/parquet_filter_pushdown.rs [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+            .with_pushdown_filters(true)
+            .with_reorder_predicates(true),
+        ParquetScanOptions::default()
+            .with_page_index(true)
+            .with_pushdown_filters(true)
+            .with_reorder_predicates(false),
+    ];
+
+    let filter_matrix = vec![
+        // Selective-ish filter
```
Comment (Ted-Jiang) [inline benchmarks/src/bin/parquet_filter_pushdown.rs:128]: well-defined test case and test data! 👍

File: benchmarks/src/bin/parquet_filter_pushdown.rs [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+    /// Number of partitions to process in parallel
+    #[structopt(long = "partitions", default_value = "2")]
+    partitions: usize,
+
+    /// Path to folder where access log file will be generated
+    #[structopt(parse(from_os_str), required = true, short = "p", long = "path")]
+    path: PathBuf,
+
+    /// Batch size when reading Parquet files
+    #[structopt(short = "s", long = "batch-size", default_value = "8192")]
```
Comment (Ted-Jiang) [inline benchmarks/src/bin/parquet_filter_pushdown.rs:74]: i think there two short options 's'

File: benchmarks/README.md [hunk 3]
```diff
@@ -126,3 +126,37 @@ h2o groupby query 1 took 1669 ms
...
 [1]: http://www.tpc.org/tpch/
 [2]: https://www1.nyc.gov/site/tlc/about/tlc-trip-record-data.page
+
+## Parquet filter pushdown benchmarks
+
+This is a set of benchmarks for testing and verifying performance of parquet filter pushdown. The queries are executed on
+a synthetic dataset generated during the benchmark execution and designed to simulate web server access logs. 
+
+```base
+cargo run --release --bin parquet_filter_pushdown --query --path ./data --scale-factor 1.0
```
Comment (Ted-Jiang) [inline benchmarks/README.md:136]: i think the `--query` is redundant

**Comment by Ted-Jiang** [comment]
@thinkharderdev thanks  for your great bench.
I run parquet tools in local get (1.0 GB)
```
(venv) yangjiang@LM-SHC-15009782 data % parquet-tools column-index  ./logs.parquet                                                     
row group 0:
column index for column service:
Boudary order: UNORDERED
                      null count  min                                       max                                     
page-0                         0  backend                                   frontend                                

offset index for column service:
                          offset   compressed size       first row index
page-0                        62               117                     0

column index for column host:
Boudary order: UNORDERED
                      null count  min                                       max                                     
page-0                         0  i-1ec3ca3151468928.ec2.internal           i-1ec408f54dbd3750.ec2.internal         

offset index for column host:
                          offset   compressed size       first row index
page-0                       566               125                     0

column index for column pod:
Boudary order: UNORDERED
                      null count  min                                       max                                     
page-0                         0  aejowuublavflbbsvlfozigwpmrxldvhaollk     zxxlzhdrucrhpicpdgxtfpyuknvviimggtq     

offset index for column pod:
                          offset   compressed size       first row index
page-0                      6689               602                     0

column index for column container:
Boudary order: UNORDERED
                      null count  min                                       max                                     
page-0                         0  backend_container_0                       frontend_container_1                    

offset index for column container:
                          offset   compressed size       first row index
page-0                      7602               593                     0
```
There are at most two pages in one col, I think if we adjust to get more pages in one col (like reduce the page size), it will get greater performance in  enable `enable_page_index `,  we can get more opportunitys to skip whole pages without decoding! 🤔

FYI,  i see impala choose to use fixed row number in one page to do benchmark for getting good performance.

File: benchmarks/src/bin/parquet_filter_pushdown.rs (hunk 2 above)
Comment (alamb) [reply benchmarks/src/bin/parquet_filter_pushdown.rs:74]: in fact when you run the example in debug mode it asserts on exactly this problem:


```shell
cargo run --bin parquet_filter_pushdown -- --path ./data --\|alamb@aal-dev:~/arrow-datafusion$ 
scale-factor 1.0
...
     Running `target/debug/parquet_filter_pushdown --path ./data --scale-factor 1.0`           |error[E0433]: failed to resolve: use of undeclared type `WriterProperties`
thread 'main' panicked at 'Argument short must be unique                                       |   --> benchmarks/src/bin/parquet_filter_pushdown.rs:235:17
                                                                                               |    |
        -s is already in use', /home/alamb/.cargo/registry/src/github.com-1ecc6299db9ec823/cla\|235 |     let props = WriterProperties::builder()
p-2.34.0/src/app/parser.rs:190:13                                                              |    |                 ^^^^^^^^^^^^^^^^ not found in this scope
note: run with `RUST_BACKTRACE=1` environment variable to display a ba
```

Comment (thinkharderdev) [reply benchmarks/src/bin/parquet_filter_pushdown.rs]: Yeah, just removed one of them. I don't think batch size needs to be a cli option in this benchmark.

**Comment by alamb** [comment]
CI check is unrelated : https://github.com/apache/arrow-datafusion/issues/3798

File: benchmarks/src/bin/parquet_filter_pushdown.rs [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+        // Nested filters
+        col("request_method").eq(lit("POST")).and(or(
+            col("response_status").eq(lit(503_u16)),
+            col("response_status").eq(lit(403_u16)),
+        )),
+        // Many filters
+        combine_filters(&[
+            col("request_method").not_eq(lit("GET")),
+            col("response_status").eq(lit(400_u16)),
+            // TODO this fails in the FilterExec with Error: Internal("The type of Dictionary(Int32, Utf8) = Utf8 of binary physical should be same")
```
Comment (alamb) [inline benchmarks/src/bin/parquet_filter_pushdown.rs:149]: Coercion!

**Pull request merged**